In [1]:
import pandas as pd

In [3]:
company_df = pd.read_csv('10-K_Data.csv')

In [4]:
company_df

Company  Fiscal Year Total Revenue(in mil) Net Income Total Assets  \
0      Tesla         2021                53,820      5,644       62,131   
1      Tesla         2022                81,462      12587       82,338   
2      Tesla         2023                96,773     14,974     1,06,618   
3  Microsoft         2021              1,68,088     61,271  \t\n333,779   
4  Microsoft         2022              1,98,270     72,738     3,64,840   
5  Microsoft         2023              2,11,915     72,361     4,11,976   
6      Apple         2021              3,65,817     94,680     3,51,002   
7      Apple         2022                394328     99,803     3,52,755   
8      Apple         2023              3,83,285     96,995     3,52,583   

  Total Liabilities CF_Operating  
0        \t\n30,548        5,644  
1        \t\n36,440       12,587  
2            43,009       14,974  
3          1,91,791       76,740  
4          1,98,298       89,035  
5          2,05,753       87,582  
6          2,87,912     1,04,038  
7          3,02,083     1,22,151  
8          2,90,437     1,10,543

In [7]:
company_df[['Total Liabilities', 'Total Assets' ]] = company_df[['Total Liabilities', 'Total Assets']].replace(to_replace=[r'\t', r'\n'], value='', regex=True)

In [8]:
company_df

Company  Fiscal Year Total Revenue(in mil) Net Income Total Assets  \
0      Tesla         2021                53,820      5,644       62,131   
1      Tesla         2022                81,462      12587       82,338   
2      Tesla         2023                96,773     14,974     1,06,618   
3  Microsoft         2021              1,68,088     61,271      333,779   
4  Microsoft         2022              1,98,270     72,738     3,64,840   
5  Microsoft         2023              2,11,915     72,361     4,11,976   
6      Apple         2021              3,65,817     94,680     3,51,002   
7      Apple         2022                394328     99,803     3,52,755   
8      Apple         2023              3,83,285     96,995     3,52,583   

  Total Liabilities CF_Operating  
0            30,548        5,644  
1            36,440       12,587  
2            43,009       14,974  
3          1,91,791       76,740  
4          1,98,298       89,035  
5          2,05,753       87,582  
6          2,87,912     1,04,038  
7          3,02,083     1,22,151  
8          2,90,437     1,10,543

In [12]:
company_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   Company                9 non-null      object
 1   Fiscal Year            9 non-null      int64 
 2   Total Revenue(in mil)  9 non-null      object
 3   Net Income             9 non-null      object
 4   Total Assets           9 non-null      object
 5   Total Liabilities      9 non-null      object
 6   CF_Operating           9 non-null      object
dtypes: int64(1), object(6)
memory usage: 636.0+ bytes


In [13]:
for column in company_df.columns[2:]:
    company_df[column] = pd.to_numeric(company_df[column].str.replace(',', '', regex=True))

## Calculate year-over-year changes for each financial metric : total revenue, net income, total assets, total liabilities and cash flows from operating activities

In [14]:
company_df['Revenue Growth (%)'] = company_df.groupby(['Company'])['Total Revenue(in mil)'].pct_change() * 100

In [15]:
company_df

Company  Fiscal Year  Total Revenue(in mil)  Net Income  Total Assets  \
0      Tesla         2021                  53820        5644         62131   
1      Tesla         2022                  81462       12587         82338   
2      Tesla         2023                  96773       14974        106618   
3  Microsoft         2021                 168088       61271        333779   
4  Microsoft         2022                 198270       72738        364840   
5  Microsoft         2023                 211915       72361        411976   
6      Apple         2021                 365817       94680        351002   
7      Apple         2022                 394328       99803        352755   
8      Apple         2023                 383285       96995        352583   

   Total Liabilities  CF_Operating  Revenue Growth (%)  
0              30548          5644                 NaN  
1              36440         12587           51.360089  
2              43009         14974           18.795267  
3             191791         76740                 NaN  
4             198298         89035           17.956071  
5             205753         87582            6.882030  
6             287912        104038                 NaN  
7             302083        122151            7.793788  
8             290437        110543           -2.800461

In [16]:
company_df['Net Income Growth (%)'] = company_df.groupby(['Company'])['Net Income'].pct_change() * 100

In [17]:
company_df

Company  Fiscal Year  Total Revenue(in mil)  Net Income  Total Assets  \
0      Tesla         2021                  53820        5644         62131   
1      Tesla         2022                  81462       12587         82338   
2      Tesla         2023                  96773       14974        106618   
3  Microsoft         2021                 168088       61271        333779   
4  Microsoft         2022                 198270       72738        364840   
5  Microsoft         2023                 211915       72361        411976   
6      Apple         2021                 365817       94680        351002   
7      Apple         2022                 394328       99803        352755   
8      Apple         2023                 383285       96995        352583   

   Total Liabilities  CF_Operating  Revenue Growth (%)  Net Income Growth (%)  
0              30548          5644                 NaN                    NaN  
1              36440         12587           51.360089             123.015592  
2              43009         14974           18.795267              18.964010  
3             191791         76740                 NaN                    NaN  
4             198298         89035           17.956071              18.715216  
5             205753         87582            6.882030              -0.518299  
6             287912        104038                 NaN                    NaN  
7             302083        122151            7.793788               5.410858  
8             290437        110543           -2.800461              -2.813543

## Growth Analysis:
Tesla saw significant growth in both revenue and net income in 2022, with revenue growing by over 51% and net income by about 123%, indicating a highly successful year with substantial improvements in sales and profitability. In 2023, growth continued but at a slower pace.

Microsoft displayed steady growth in revenue and net income from 2021 to 2022. However, in 2023, while revenue grew modestly by 6.88%, net income slightly decreased by 0.52%, which could be due to increased costs, changes in the market, or other external factors affecting profitability.

Apple showed a consistent revenue increase from 2021 to 2022 but experienced a small decline in both revenue (-2.80%) and net income (-2.81%) in 2023. This decline could indicate market saturation, increased competition, or other operational challenges.

##  Summary by Company across all years

In [18]:
summary_by_company = company_df.groupby('Company').agg({
    'Total Revenue(in mil)': 'sum',
    'Net Income': 'sum',
    'Total Assets': 'sum',
    'Total Liabilities': 'sum',
    'CF_Operating': 'sum'
})

In [19]:
summary_by_company

Total Revenue(in mil)  Net Income  Total Assets  Total Liabilities  \
Company                                                                         
Apple                    1143430      291478       1056340             880432   
Microsoft                 578273      206370       1110595             595842   
Tesla                     232055       33205        251087             109997   

           CF_Operating  
Company                  
Apple            336732  
Microsoft        253357  
Tesla             33205

## Insights:
Market Scale and Growth:

* Apple shows the highest revenue and net income, highlighting its dominant market position and operational efficiency.
* Microsoft, while having less revenue than Apple, shows substantial total assets, which might be indicative of significant investments in technology, infrastructure, or acquisitions.
* Tesla, despite having the lowest figures in most categories, shows remarkable growth potential, given its revenue and net income relative to its younger market tenure.
  
Profitability and Efficiency:

* Apple's high net income relative to revenue suggests strong profitability and effective cost management.
* Microsoft also displays strong profitability but with a higher asset base, indicating potentially lower asset turnover.
* Tesla’s net income, while lower, aligns with its growth stage and heavy reinvestment in expansion and innovation.

Financial Health and Stability:

* Asset to Liability Ratio: Apple and Microsoft show healthy ratios, suggesting good financial stability and lower risk. Tesla’s ratio indicates higher financial risk but is not unusual for a rapidly growing company in a capital-intensive industry.
* Operational Cash Flow: High figures for Apple and Microsoft confirm strong operational health and cash management. Tesla’s operational cash flow matching its net income may suggest heavy cash reinvestment needs.

## Summary by Year across all companies

In [20]:
summary_by_year = company_df.groupby('Fiscal Year').agg({
    'Total Revenue(in mil)': 'sum',
    'Net Income': 'sum',
    'Total Assets': 'sum',
    'Total Liabilities': 'sum',
    'CF_Operating': 'sum'
})

In [21]:
summary_by_year

Total Revenue(in mil)  Net Income  Total Assets  \
Fiscal Year                                                    
2021                        587725      161595        746912   
2022                        674060      185128        799933   
2023                        691973      184330        871177   

             Total Liabilities  CF_Operating  
Fiscal Year                                   
2021                    510251        186422  
2022                    536821        223773  
2023                    539199        213099